### ISSUE TITLE
Add mutable filed in `Object expression`

### ISSUE THREAD

**Issue by greatim** [fsharp/fslang-suggestions#946]
# Add mutable filed in `Object expression`

Currently when we use **object expressions**, the outer mutable variable will be promoted to `Ref`, if multiple variable variables are used, multiple `Ref`s will be generated. In fact, most of the time, we don’t share these variables, can a method to add variable variables directly in the object is provided?
For example, in the following code
```
  let foo() = 
      let mutable isDisposed = false
      {
         new IDisposable with
            member __.Dispose() = 
               if not isDisposed then
                  isDisposed <- true
      }
```
can the variable `isDisposed` be moved inside

```
  let foo() =       
      {
         let mutable isDisposed = false
         new IDisposable with
            member __.Dispose() = 
               if not isDisposed then
                  isDisposed <- true
      }
```
or can use other alternatives?

**Comment by Happypig375** [comment]
I believe that a more sensible approach would be to allow `let` and `member val` bindings in object expressions just like in classes, with the same syntax.
```fs
  let foo() =       
      {
         new IDisposable with
            let mutable isDisposed = false
            member val K = 1
            member __.Dispose() = 
               if not isDisposed then
                  isDisposed <- true
      }
```

**Comment by cartermp** [comment]
Promoting the mutable value to a `ref` when it's closed over in an object expression like that is by design. Object expressions are objects that are allocated on the heap, so we need to promote this to the heap. It's just like when we promote a mutable value to a `ref` when it is passed to a lambda.

I agree with @Happypig375 that any work in this area would be to allow scoping mutability to the expression itself. Is that what you'd like to suggest?

**Comment by greatim** [comment]
> th @Happypig375 that

Yes, I think, the object generated by the object expression has already been allocated in the heap, if it is not shared, there is no need to generate another object

**Comment by greatim** [comment]
> I believe that a more sensible approach would be to allow `let` and `member val` bindings in object expressions just like in classes, with the same syntax.
> 
> ```fsharp
>   let foo() =       
>       {
>          new IDisposable with
>             let mutable isDisposed = false
>             member val K = 1
>             member __.Dispose() = 
>                if not isDisposed then
>                   isDisposed <- true
>       }
> ```

If it is written below, it is necessary to distinguish between `member` and `override`. Currently, `member` and `override` are the same, And if you use `object + interface`, you need to consider more permissions, as I mentioned in another [discussion](https://github.com/fsharp/fslang-suggestions/issues/947)

**Comment by dsyme** [comment]
I agree `let mutable` should be usable inside the scope of the object expression

Marking this as approved in principle.
